user feature install fails with Gradle 7

Running `gradle installFeature` to install user feature fails on Gradle 7.

Related Gradle issue:
https://github.com/gradle/gradle/issues/13122

Problem happens when we recompile Gradle using the ProjectBuilder API to fetch new features.json artifact (generated by prepareFeature) from the Mavenlocal repo. 

==Error log & stack trace==
* What went wrong:
Execution failed for task ':installFeature'.
> The settings are not yet available for build 'regFeature'.
...
Caused by: java.lang.IllegalStateException: The settings are not yet available for build 'regFeature'.
        at org.gradle.invocation.DefaultGradle.getSettings(DefaultGradle.java:207)
        at org.gradle.invocation.DefaultGradle_Decorated.getSettings(Unknown Source)
        at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.ignoresSettingsRepositories(DefaultConfiguration.java:756)

--- a/src/main/groovy/io/openliberty/tools/gradle/tasks/AbstractFeatureTask.groovy
+++ b/src/main/groovy/io/openliberty/tools/gradle/tasks/AbstractFeatureTask.groovy
@@ -131,7 +131,7 @@ public class AbstractFeatureTask extends AbstractServerTask {
         public File downloadArtifact(String groupId, String artifactId, String type, String version) throws PluginExecutionException {
  			//check if jsonCoordinate is provided from prepareFeature task
 			def coordinates = groupId + ":" + artifactId + ":" + version
- 			if(jsonCoordinate && jsonCoordinate.equals(coordinates)) {
+ 			if(jsonCoordinate != null && !jsonCoordinate.isEmpty() && jsonCoordinate.equals(coordinates)) {
  				Configuration provided = project.getConfigurations().findByName("jsonProvided");
  				if (provided != null) {
  					return provided.getFiles().iterator().next()

--- a/src/main/groovy/io/openliberty/tools/gradle/tasks/AbstractFeatureTask.groovy
+++ b/src/main/groovy/io/openliberty/tools/gradle/tasks/AbstractFeatureTask.groovy
@@ -27,6 +27,7 @@ import org.gradle.api.logging.LogLevel
 import org.gradle.api.tasks.Internal
 import org.gradle.api.tasks.options.Option
 import org.gradle.testfixtures.ProjectBuilder
+import org.gradle.api.artifacts.Configuration
 
 public class AbstractFeatureTask extends AbstractServerTask {
 
@@ -38,10 +39,10 @@ public class AbstractFeatureTask extends AbstractServerTask {
 
     private InstallFeatureUtil util;
 
-    @Internal
-    Project newProject = project;
-
     private ServerFeatureUtil servUtil;
+	
+	@Internal
+ 	String jsonCoordinate;
 
     @Option(option = 'serverDir', description = '(Optional) Server directory to get the list of features from.')
     void setServerDirectoryParam(String serverDir) {
@@ -128,8 +129,18 @@ public class AbstractFeatureTask extends AbstractServerTask {
 
         @Override
         public File downloadArtifact(String groupId, String artifactId, String type, String version) throws PluginExecutionException {
-            return ArtifactDownloadUtil.downloadArtifact(project, groupId, artifactId, type, version);
-        }
+ 			//check if jsonCoordinate is provided from prepareFeature task
+			def coordinates = groupId + ":" + artifactId + ":" + version
+ 			if(jsonCoordinate && jsonCoordinate.equals(coordinates)) {
+ 				Configuration provided = project.getConfigurations().findByName("jsonProvided");
+ 				if (provided != null) {
+ 					return provided.getFiles().iterator().next()
+ 				}else {
+ 					return ArtifactDownloadUtil.downloadArtifact(project, groupId, artifactId, type, version);
+ 				}
+ 			}
+ 			return ArtifactDownloadUtil.downloadArtifact(project, groupId, artifactId, type, version);
+ 		}
     }
 
     protected Set<String> getPluginListedFeatures(boolean findEsaFiles) {
@@ -160,6 +171,7 @@ public class AbstractFeatureTask extends AbstractServerTask {
         List<String> result = new ArrayList<String>()
         project.configurations.featuresBom.dependencies.each { dep ->
             def coordinate = dep.group + ":" + "features" + ":" + dep.version
+			jsonCoordinate = coordinate
             logger.debug("feature Json: " + coordinate)
             result.add(coordinate)
         }
@@ -232,23 +244,6 @@ public class AbstractFeatureTask extends AbstractServerTask {
     }
 
     protected InstallFeatureUtil getInstallFeatureUtil(Set<String> pluginListedEsas, List<ProductProperties> propertiesList, String openLibertyVerion, String containerName, List<String> additionalJsons) throws PluginExecutionException {
-        //if installing userFeature, recompile gradle to find mavenLocal artifacts created by prepareFeature task.
-        if (project.configurations.featuresBom.dependencies) {
-            try {
-                ProjectBuilder builder = ProjectBuilder.builder();
-                newProject = builder
-                        .withProjectDir(project.rootDir)
-                        .withGradleUserHomeDir(project.gradle.gradleUserHomeDir)
-                        .withName(project.name)
-                        .build();
-
-                // need this for gradle to evaluate the project
-                // and load the different plugins and extensions
-                newProject.evaluate();
-            } catch (Exception e) {
-                throw new PluginExecutionException("Could not parse build.gradle " + e.getMessage());
-            }
-        }
         createNewInstallFeatureUtil(pluginListedEsas, propertiesList, openLibertyVerion, containerName, additionalJsons)
         return util
     }

--- a/src/main/groovy/io/openliberty/tools/gradle/tasks/AbstractPrepareTask.groovy
+++ b/src/main/groovy/io/openliberty/tools/gradle/tasks/AbstractPrepareTask.groovy
@@ -23,10 +23,12 @@ import org.gradle.api.logging.LogLevel
 import org.gradle.api.tasks.TaskAction
 import org.gradle.api.tasks.options.Option
 import org.gradle.api.tasks.Internal
+import org.gradle.api.artifacts.dsl.DependencyHandler
 
 import io.openliberty.tools.common.plugins.util.PrepareFeatureUtil
 import io.openliberty.tools.common.plugins.util.PluginExecutionException
 import io.openliberty.tools.common.plugins.util.PluginScenarioException
+import io.openliberty.tools.gradle.utils.ArtifactDownloadUtil
 
 public class AbstractPrepareTask extends AbstractServerTask {
 
@@ -81,38 +83,16 @@ public class AbstractPrepareTask extends AbstractServerTask {
 
         @Override
         public File downloadArtifact(String groupId, String artifactId, String type, String version) throws PluginExecutionException {
-            String coordinates = groupId + ":" + artifactId + ":" + version + "@" + type
-            def dep = project.dependencies.create(coordinates)
-            def config = project.configurations.detachedConfiguration(dep)
-
-            Set<File> files = new HashSet<File>()
-            try {
-                config.resolvedConfiguration.resolvedArtifacts.each { artifact ->
-					File artifactFile = artifact.file
-					if(artifactFile.absolutePath.contains(groupId)){
-						//loaded from remote repository
-						String mavenLocalRepo = System.getenv("FEATURE_LOCAL_REPO")
-						if(!mavenLocalRepo) {
-							mavenLocalRepo = new File(System.getProperty("user.home")+ "/.m2/repository")
-						}
-						groupId = groupId.replace(".", "/");
-						String artifactPath = String.format("%s/%s/%s/%s-%s.%s", groupId, artifactId, version, artifactId, version,type);
-						File artifactMavenLocal = new File(mavenLocalRepo + File.separator + artifactPath)
-						FileUtils.copyFile(artifactFile, artifactMavenLocal) //copy downloaded artifacts from gradle cache to maven local repo
-						files.add(artifactMavenLocal)
-					}else {
-						files.add(artifactFile)
-					}
-                }
-            } catch (ResolveException e) {
-                throw new PluginExecutionException("Could not find artifact with coordinates " + coordinates)  
-            }
-
-            if (!files) {
-                throw new ResolveException("Could not find artifact with coordinates " + coordinates)
-            }
-            return files.iterator().next()
+			return ArtifactDownloadUtil.downloadArtifact(project, groupId, artifactId, type, version);
         }
+		
+		@Override
+		public void provideJsonFileDependency(File file) {
+			project.getConfigurations().create('jsonProvided')
+			DependencyHandler dependencies = project.getDependencies();
+			dependencies.add("jsonProvided", project.files(file))
+		}
+	
     }
 	
 	@Internal

